### ISSUE TITLE
Add proper ACPI support

### ISSUE THREAD

**Issue by 9xbt** [CosmosOS/Cosmos#2503]
I want to detect when the user presses the phisical power button on the PC and save unsaved progress to disk and then shut down and it seems like that feature is not implemented. Can Cosmos devs implement it? Thanks

**Comment by zarlo** [comment]
https://github.com/CosmosOS/Cosmos/blob/master/source/Cosmos.Core/ACPI.cs

**Comment by AsertCreator** [comment]
is it possible to detect cosmos running on laptop and read battery status (charging, discharging, percentage). i think it is possible to implement

**Comment by zarlo** [comment]
yes that can be implemented

**Comment by zarlo** [comment]
i would wait for https://github.com/CosmosOS/Cosmos/pull/2288 but

**Comment by 9xbt** [comment]
okay yeah nice but how can i detect when the pc wants to shut down @zarlo

**Comment by AsertCreator** [comment]
oh i know! we need to override power functions. better rewrite entire power class with ability to stop shutdown/restart

**Comment by AsertCreator** [comment]
but i dont know if we can detect acpi button rn

**Comment by 9xbt** [comment]
f

**Comment by MishaTY** [comment]
Basic ACPI support already existed for a long time, however we need an ACPI interrupter and to handle the SCI.

**Comment by 9xbt** [comment]
i know

**Comment by MishaProductions** [comment]
Nope, still need to figure out how to port LAI correctly as my port has stack issues

**Comment by ADev531** [comment]
ACPI has functioon like detect cpu temp, etc. i think we can get apic.

**Comment by zarlo** [comment]
>  we can use usb mouse or keyboard?

it will help us make drivers for it as it will give use access to better timers

**Comment by ADev531** [comment]
Is there an updates?

**Comment by valentinbreiz** [comment]
> Is there an updates?

Yes LAI is being ported to Cosmos

**Comment by ADev531** [comment]
is there update again? its going great, i think

**Comment by ADev531** [comment]
Will this stop and continue at gen 3?

**Comment by valentinbreiz** [comment]
Cosmos Gen3 has replaced the IL2CPU-based Gen2 codebase, and `gen3` is now the default branch of this repository. This issue was filed against Gen2, which is no longer developed, so it is being closed as part of the Gen3 cutover (#3088).

If you still hit this on Gen3, please open a fresh issue. The compiler, the build pipeline and the debugging story are all different now, so a new report will get a far more useful answer than reviving this one. Start here: https://cosmosos.github.io/

Gen2 remains on the `gen2` branch, and this issue stays readable.
